pkg/cli/sanitize/hide_secrets.go

```diff
@@ -0,0 +1,207 @@
+/*
+Copyright The Helm Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package sanitize
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/pkg/errors"
+
+	"helm.sh/helm/v3/pkg/release"
+
+	"gopkg.in/yaml.v3"
+)
+
+const (
+	hiddenSecretValue = "[HIDDEN]"
+)
+
+// HideManifestSecrets sanitizes release manifest and replaces it.
+// Manifest cannot be applied after this operation.
+// Indentation and extra spaces in Secret's `data` and `stringData` sections can be impacted.
+func HideManifestSecrets(r *release.Release) error {
+	if r == nil {
+		return nil
+	}
+	manifest, err := hideSecrets(r.Manifest)
+	if err != nil {
+		return err
+	}
+
+	r.Manifest = manifest
+
+	return nil
+}
+
+// hideSecrets replaces values in Secrets in the chart manifest with
+// `[HIDDEN]` value.
+func hideSecrets(manifest string) (string, error) {
+	resources := strings.Split(manifest, "\n---")
+	outRes := make([]string, 0, len(resources))
+
+	for _, r := range resources {
+		var resourceMap map[string]interface{}
+		err := yaml.Unmarshal([]byte(r), &resourceMap)
+		if err != nil {
+			return "", errors.Wrapf(err, "failed to unmarshal %q resource", tryToGetName(resourceMap))
+		}
+
+		if isSecret(resourceMap) {
+			rs, err := hideSecretData(r)
+			if err != nil {
+				return "", errors.Wrapf(err, "failed to hide %q Secret data", tryToGetName(resourceMap))
+			}
+			r = rs
+		}
+
+		outRes = append(outRes, r)
+	}
+
+	return strings.Join(outRes, "\n---"), nil
+}
+
+func tryToGetName(resourceMap map[string]interface{}) string {
+	metadata, ok := resourceMap["metadata"].(map[string]interface{})
+	if !ok {
+		return ""
+	}
+
+	name, ok := metadata["name"].(string)
+	if !ok {
+		return ""
+	}
+
+	return name
+}
+
+func isSecret(resource map[string]interface{}) bool {
+	kind, ok := resource["kind"].(string)
```

Comment:
A MUCH cleaner way to handle all of this validation logic would be to convert to a `runtime.Object`, then check the value of its `.GroupVersionKind()`.

Have a look here for an example: https://github.com/helm/helm/blob/e9e6b07f62e2bb5067afe79782ba3dd85efbd00e/pkg/kube/client.go#L594

https://pkg.go.dev/k8s.io/apimachinery@v0.21.3/pkg/runtime

https://pkg.go.dev/k8s.io/apimachinery@v0.21.3/pkg/runtime/schema#GroupVersionKind